Workaround for wrong TemplateType.

jDramaix opened the pull request:
Closure compiler adds wrong TemplateType (like IObject#KEY or IObject#Value) for sereval types. The goal of this change is to detect and skip these TemplateType.

jDramaix committed:
Workaround for wrong TemplateType.

Closure compiler adds wrong TemplateType (like IObject#KEY or IObject#Value) for sereval types.
The goal of this change is to detect and skip these TemplateType.

jDramaix committed:
Upgrade closure-compiler to the last version.

- Removes an old option skipParseExterns that was not used by tests.
- filters out extern symbols on the window object that are identical
  to already processed symbols.
- adds es6_min.js that contains the minimum of es6.js needed for
  processing input. This is only used for testing.

jDramaix committed:
Workaround for wrong TemplateType.

Closure compiler adds wrong TemplateType (like IObject#KEY or IObject#Value) for sereval types. The goal of this change is to detect and skip these TemplateType.

rkirov commented on src/main/java/com/google/javascript/clutz/DeclarationGenerator.java:
```diff
@@ -1116,15 +1117,25 @@ private void visitVarDeclaration(String name, JSType type) {
+        for (TemplateType templateType: type.getTemplateTypeMap().getTemplateKeys()) {
+          String displayName = templateType.getDisplayName();
+
+          if (displayName.indexOf('#') < 0) {
```
I have run into this issue before, I think it is safe to just strip the part before `#` , instead of complete omitting it.
 So the TS type will become `Object<Key, Value>`.

jDramaix replied:
But Object constructor function doesn't define any template type. IObject does. Same with Array, if we strip the part before the # instead of omitting it, we generate `class Array<VALUE1, T>` instead of `class Array<T>`

rkirov replied:
Good point, my latest thinking is that `IObject#` type arguments should be skipped for all but `IObject` itself.

I made some changes onto your PR and opened a main repo PR here https://github.com/angular/clutz/pull/280. Added a few more tests too. Pls, review it there.

jDramaix committed:
Best management of TemplateType.

- Closure compiler adds weird TemplateType (like IObject#KEY or IObject#Value) for several types that doesn;'t exist in the type definition. The goal of this change is to detect and skip these TemplateType.
- Don't redefine type parameter at method level if the same type parameter is defined on the declaring Type.

jDramaix committed:
Merge branch 'jd_template_workaround' of https://github.com/jDramaix/clutz into jd_template_workaround
